lab3

OkDenAl opened the pull request:
lab3
Status: done

OkDenAl committed:
init branch

OkDenAl committed:
add first lstar implementation

OkDenAl committed:
add isEqual check to MATMock

OkDenAl committed:
change main.cpp and start to implement main algo

OkDenAl committed:
add prefix and suffix languages filling

OkDenAl committed:
fix bug and add fix states for deleting traps

OkDenAl committed:
add circuit-finding-algorithm

wizardOfOz21 committed:
Добавлена функция поиска путей в автомате

wizardOfOz21 committed:
Исправлен баг

OkDenAl committed:
add letters paths

OkDenAl committed:
fix

wizardOfOz21 committed:
Главный алгоритм

wizardOfOz21 committed:
Merge branch 'lab3' of https://github.com/wizardOfOz21/tfl_labs into lab3

wizardOfOz21 committed:
Исправление алгоритма поиска слов вдоль путей

wizardOfOz21 committed:
Добавлено сопоставление накачек

wizardOfOz21 committed:
Изменение алгоритма поиска циклов

OkDenAl committed:
add readme

OkDenAl committed:
change main

OkDenAl committed:
add our code from lab2 for parsing, change oracle defining

wizardOfOz21 committed:
Исправление бага с поиском циклов

Объединять поиск циклов и поиск обычных путей - плохая идея

wizardOfOz21 committed:
В MATMock добавлен основной язык

wizardOfOz21 committed:
Проверка накачек

OkDenAl committed:
Update README.md

OkDenAl committed:
Update README.md

OkDenAl committed:
main.cpp uncomented

TonitaN commented:
Понятно, что решение с регулярками временное, а на КС-языках вы тестировали решение? В принципе, в регулярных языках все накачки, если уж они есть, совместны друг с другом.

TonitaN commented:
У меня нет результата (похоже, утечка) на тесте:
```
    std::string mainLangRegex="^(aa)*a(bb)*(c|a)$";
    std::string suffixLangRegex="^((aa)*a?a(bb)*(c|a)|c|a|(bb)*b?(c|a))$";
```
Параметры `-a abc -ra 15 -ml 15 -pi 15`

Запускала в режиме release. Проверьте, пожалуйста, воспроизводится ли у вас это поведение.

OkDenAl committed:
add CFL examples

OkDenAl commented:
> Понятно, что решение с регулярками временное, а на КС-языках вы тестировали решение? В принципе, в регулярных языках все накачки, если уж они есть, совместны друг с другом.

добавили примеры КС языков для тестирования. парсер в процессе фикса :)

wizardOfOz21 committed:
Добавлено замечание про язык Дика

wizardOfOz21 committed:
Правки нейминга и обертки парсера

- Обертка парсера теперь бросает специальное исключение, которое ловится при её использовании
- Изменена генерация автомата для языка суффиксов

wizardOfOz21 committed:
Указатели заменены на ссылки

wizardOfOz21 committed:
Раскоментирован прием параметров

wizardOfOz21 commented:
> У меня нет результата (похоже, утечка) на тесте:
> 
> ```
>     std::string mainLangRegex="^(aa)*a(bb)*(c|a)$";
>     std::string suffixLangRegex="^((aa)*a?a(bb)*(c|a)|c|a|(bb)*b?(c|a))$";
> ```
> 
> Параметры `-a abc -ra 15 -ml 15 -pi 15`
> 
> Запускала в режиме release. Проверьте, пожалуйста, воспроизводится ли у вас это поведение.

Парсер не знает о вопросиках, но не падает из-за этого. Дело было в непойманном исключении, которое бросала обертка над ним при неудачном результате парсинга, теперь оно ловится.

TonitaN commented:
Так, а всё ли корректно с примером Дика? Некоторые накачки имеют вид типа таких:
```
aaa(c)*bbbbbb(b)*
```
Тут вообще непонятно, как слово оказалось в языке, т.к. `b`-закрывающих явно больше, чем открывающих `a`. Или нужно прояснить, что напечаталось в выводе, если это не целая накачка, а только её половина.

OkDenAl committed:
hot fix

OkDenAl committed:
Merge branch 'lab3' of github.com:wizardOfOz21/tfl_labs into lab3

OkDenAl commented:
> Так, а всё ли корректно с примером Дика? Некоторые накачки имеют вид типа таких:
> 
> ```
> aaa(c)*bbbbbb(b)*
> ```
> 
> Тут вообще непонятно, как слово оказалось в языке, т.к. `b`-закрывающих явно больше, чем открывающих `a`. Или нужно прояснить, что напечаталось в выводе, если это не целая накачка, а только её половина.

в процессе выполнения лабораторки переезжали с одних тегов командной строки на другие, не уследили за сменой тегов, поэтому в make файле были неправильные аргументы командной строки :( из-за этого была такая ситуация с языком Дика
сейчас выпустил фикс

TonitaN commented:
А сколько примерно ждать окончания разбора языка Дика при следующих значениях ключей?
```
-a abc -ra 15 -ml 15 -pi 15
```
(у меня в процессе ещё несколько лаб проверяется, поэтому работает ожидаемо медленнее, чем будет работать у вас)

OkDenAl commented:
> А сколько примерно ждать окончания разбора языка Дика при следующих значениях ключей?
> 
> ```
> -a abc -ra 15 -ml 15 -pi 15
> ```
> 
> (у меня в процессе ещё несколько лаб проверяется, поэтому работает ожидаемо медленнее, чем будет работать у вас)

к сожалению, опять таки проблемы нашей невнимательности в Makefile в предыдущей версии сказываются
валидные значения ключей:
```
-a abc -ra 15 -mtc 15 -pi 15
```
для более полных результатов рекомендуем брать `-mtc 1000` и выше, т.к при малых значениях контрпримеры возвращаются короткие, а могут и не вернуться вообще в некоторых кейсах, поэтому автоматы префиксов и суффиксов более скудные выходят

работает достаточно быстро

TonitaN commented:
Да, заработало и с этим оракулом адекватно.
С регулярками возник вопрос: есть ли шанс домучить до нахождения накачки, содержащей как буквы `a`, так и буквы `b`, при вот таком входе?
```
    std::string langRegex = "^bb(b*ab*aa)*$";
    std::string reverseRegex  = "^(aab*ab*)*bb$";

```
Я думала, что проблема в ключе `ra`, но даже его увеличение до 50 не помогает их найти.

OkDenAl commented:
> Да, заработало и с этим оракулом адекватно. С регулярками возник вопрос: есть ли шанс домучить до нахождения накачки, содержащей как буквы `a`, так и буквы `b`, при вот таком входе?
> 
> ```
>     std::string langRegex = "^bb(b*ab*aa)*$";
>     std::string reverseRegex  = "^(aab*ab*)*bb$";
> ```
> 
> Я думала, что проблема в ключе `ra`, но даже его увеличение до 50 не помогает их найти.

нет, он будет содержать либо a либо b
согласно условию у нас не учитываются одиночные внутренние циклы 

вот граф для префиксов в abc/ab 
![image](https://github.com/wizardOfOz21/tfl_labs/assets/94537822/112c6dfc-4850-4983-94a9-0136d8cdcdb9)

вот граф для суффиксов в abc/ab
![image](https://github.com/wizardOfOz21/tfl_labs/assets/94537822/bcd6b483-da61-4da4-b714-f282409d4916)

TonitaN commented:
Условие специфическое, но с его учётом накачки регулярок строятся корректно. По другим языкам ваши оракулы ок, свой писать сейчас не буду (сэкономлю на компромат). Зачёт.
